plugins/trace_api_plugin/include/eosio/trace_api_plugin/request_handler.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <fc/variant.hpp>
+#include <eosio/trace_api_plugin/metadata_log.hpp>
+#include <eosio/trace_api_plugin/data_log.hpp>
+
+namespace eosio::trace_api_plugin {
+
+   using now_function = std::function<fc::time_point()>;
+
+   namespace detail {
+      class response_formatter {
+      public:
+         static fc::variant process_block( const block_trace_v0& trace, const  std::optional<lib_entry_v0>& best_lib, const now_function& now, const fc::time_point& deadline );
+      };
+   }
+
+   class deadline_exceeded : public std::runtime_error {
+   public:
+      explicit deadline_exceeded(const char* what_arg)
+      :std::runtime_error(what_arg)
+      {}
+      explicit deadline_exceeded(const std::string& what_arg)
+            :std::runtime_error(what_arg)
+      {}
+   };
+
+   class bad_data_exception : public std::runtime_error {
+   public:
+      explicit bad_data_exception(const char* what_arg)
+      :std::runtime_error(what_arg)
+      {}
+
+      explicit bad_data_exception(const std::string& what_arg)
+      :std::runtime_error(what_arg)
+      {}
+   };
+
+   template<typename LogfileProvider>
+   class request_handler {
+   public:
+      request_handler(LogfileProvider&& logfile_provider, const now_function& now)
+      :logfile_provider(logfile_provider)
+      ,now(now)
+      {
+      }
+
+      /**
+       * Fetch the trace for a given block height and convert it to a fc::variant for conversion to a final format
+       * (eg JSON)
+       *
+       * @param block_height - the height of the block whose trace is requested
+       * @param deadline - deadline for the processing of the request
+       * @return a properly formatted variant representing the trace for the given block height if it exists, an
+       * empty variant otherwise.
+       * @throws deadline_exceeded when the processing of the request cannot be completed by the provided time.
+       */
+      fc::variant get_block_trace( uint32_t block_height, const fc::time_point& deadline ) {
+         std::optional<uint64_t> block_offset;
+         std::optional<lib_entry_v0> best_lib;
+         bool deadline_past = false;
+
+         // scan for the block offset and latest LIB
+         try {
+            logfile_provider.scan_metadata_log_from(block_height, 0, [&](const metadata_log_entry& e) -> bool {
+               if (e.contains<block_entry_v0>()) {
+                  const auto& block = e.get<block_entry_v0>();
+                  if (block.number == block_height) {
+                     block_offset = block.offset;
+                  }
+               } else if (e.contains<lib_entry_v0>()) {
```

Comment:
I am not sure how scan_metadata_log_from would return a lib_entry_v0 based on just a block_height value?  I was thinking the lib_entry_v0 was just for the readers understanding of the file data, but now thinking we need to work through the use case that causes the reader to return it.

Reply:
The logfile reader should just be reading sequentially with no inspection on what its passing.   As long as the slice file for a given block height contains all `lib_entry_v0` structures relevant to blocks in that slice then we should be able to find the "Best" one from that scan and it should ideally be an entry which is at or past the end of the slice once that slice is final.

Reply:
Will accept the PR and move this Discussion off comments

Reply:
ignore me, forgot that my reader just keeps reading till you tell it to stop.